pkg/volume/util/operationexecutor/operation_executor.go

```diff
@@ -0,0 +1,763 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package operationexecutor implements interfaces that enable execution of
+// attach, detach, mount, and unmount operations with a goroutinemap so that
+// more than one operation is never triggered on the same volume.
+package operationexecutor
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/util/goroutinemap"
+	"k8s.io/kubernetes/pkg/volume"
+	volumetypes "k8s.io/kubernetes/pkg/volume/util/types"
+)
+
+// OperationExecutor defines a set of operations for attaching, detaching,
+// mounting, or unmounting a volume that are executed with a goroutinemap which
+// prevents more than one operation from being triggered on the same volume.
+//
+// These operations should be idempotent (for example, AttachVolume should
+// still succeed if the volume is already attached to the node, etc.). However,
+// they depend on the volume plugins to implement this behavior.
+//
+// Once an operation completes successfully, the actualStateOfWorld is updated
+// to indicate the volume is attached/detached/mounted/unmounted.
+//
+// If the OperationExeuctor fails to start the operation, because, for
+// example, an operation with the same UniqueVolumeName is already pending, a
+// non-nil error is returned.
+//
+// Once the operation is started, since it is executed asynchronously,
+// errors are simply logged and the goroutine is terminated without updating
+// actualStateOfWorld (callers are responsible for retrying as needed).
+type OperationExecutor interface {
+	// AttachVolume spawns a new goroutine to execute volume-specific logic to
+	// attach the volume to the node specified in the volumeToAttach.
+	AttachVolume(volumeToAttach VolumeToAttach, actualStateOfWorld ActualStateOfWorldAttacherUpdater) error
+
+	// DetachVolume spawns a new goroutine to execute volume-specific logic to
+	// detach the volume from the node specified in volumeToDetach.
+	DetachVolume(volumeToDetach AttachedVolume, actualStateOfWorld ActualStateOfWorldAttacherUpdater) error
+
+	// MountVolume spawns a new goroutine to execute volume-specific logic to
+	// mount the volume to the pod specified in the volumeToMount.
+	MountVolume(waitForAttachTimeout time.Duration, volumeToMount VolumeToMount, actualStateOfWorld ActualStateOfWorldMounterUpdater) error
+
+	// UnmountVolume spawns a new goroutine to execute volume-specific logic to
```

Comment:
I think the new goroutine language is a little specific for an interface declaration.  Scratch that part on all of these.